src/browser/provider/pool.js

```diff
@@ -0,0 +1,153 @@
+import { assignIn } from 'lodash';
+import Promise from 'pinkie';
+import { PathBrowserProvider, LocalBrowserProvider, RemoteBrowserProvider } from './built-in-providers';
+import SaucelabsBrowserProvider from './saucelabs';
+import PhantomJSProvider from './phantomjs';
+import BrowserProviderCacheItem from './cache-item';
+import BrowserConnection from '../connection';
+import { GeneralError } from '../../errors/runtime';
+import MESSAGE from '../../errors/runtime/message';
+
+
+const BROWSER_PROVIDER_RE = /^([^:]+)(?::(.*))?$/;
+
+const builtInProviders = {
+    'local':     LocalBrowserProvider,
+    'path':      PathBrowserProvider,
+    'remote':    RemoteBrowserProvider,
+    'saucelabs': SaucelabsBrowserProvider,
+    'phantomjs': PhantomJSProvider
+};
+
+
+var providersCache = {};
+
+
+export default {
+    async _handlePathAndCmd (browserDesc) {
+        var providerName = 'path';
+        var provider     = await this.getProvider(providerName);
+        var options      = browserDesc.cmd ? { cmd: browserDesc.cmd } : null;
+
+        return { provider, providerName, aliases: [browserDesc.path], options };
+    },
+
+    async _parseAlias (browserDesc) {
+        var providerRegExpMatch = BROWSER_PROVIDER_RE.exec(browserDesc);
+
+        if (!providerRegExpMatch)
+            throw new GeneralError(MESSAGE.cantFindBrowser, browserDesc);
+
+        var providerName = '';
+        var alias        = '';
+
+        var provider = await this.getProvider(providerRegExpMatch[1]);
+
+        if (provider) {
+            providerName = providerRegExpMatch[1];
+            alias        = providerRegExpMatch[2];
+        }
+        else {
+            providerName = 'local';
+            provider     = await this.getProvider(providerName);
+            alias        = providerRegExpMatch[1];
+        }
+
+        if (alias === 'all') {
```

Comment:
Yeah, just figured out that provider can work only with one browser at a time. 

> What if someone will implement a custom browser provider for a company internal CI, and they want to run tests in all available browsers?

They will name it `{mycompany}-all`